### PR TITLE
einsum implementation

### PR DESCRIPTION
**Pull request by sjperkins**
- [x] Tests added / passed
- [x] Passes `flake8 dask`
- [x] Fully documented, including `docs/source/changelog.rst` for all changes
      and one of the `docs/source/*-api.rst` files for new API

### CODE REVIEW

**Comment by sjperkins** [comment]
Based on the implementation by @chrisbarber. My first two test cases worked easily and I hoped it would be trivial but there's something going wrong with the third one, based on one of opt_einsum's doc examples.

Putting this up here as a WIP and for input from people who know einsum better than I do!

/cc @dgasmith and @jcmgray

**Comment by mrocklin** [comment]
cc @shoyer @jakirkham  you may want to see this

File: dask/array/tests/test_routines.py [hunk 1]
```diff
@@ -1286,3 +1286,54 @@ def test_result_type():
     # dask 0d-arrays are NOT treated like scalars
     c = da.from_array(np.ones((), np.float64), chunks=())
     assert da.result_type(a, c) == np.float64
+
+
+def test_einsum():
```
Comment (mrocklin) [inline dask/array/tests/test_routines.py]: Einsum seems tricky enough that I suggest we consider building an additional test that is parametrized over many different kinds of situations:

Perhaps something like the following (untested):
```python
@pytest.mark.parametrize('signature', [
    'abc,bad->abcd', 
    'ab,bc-> ...' 
    ...
])
def test_einsum(signature):
    dimensions = {'a': 5, 'b': 6, 'c': 7, 'd': 5, 'e': 6}

    def shape_from_string(s):
        return tuple(dimensions[c] for c in s)

    np_inputs = [np.random.random(shape_from_string(s) for s in signature.split('->')[0].split(',')]
    da_inputs = [da.from_array(x, chunks=4) for x in np_inputs]

    assert_eq(np.einsum(signature, *np_inputs), 
              da.einsum(signature, *da_inputs))
```

**Comment by shoyer** [comment]
I think it would be straightforward to only support `optimize` if using NumPy 1.12+. We could set a default value of `optimize=None` which means "optimize if possible."

A comment on the implementation: it would be best to avoid using `concatenate=True`, which means that all chunks corresponding to contracted dimensions get concatenated together. In practice, this means that `concatenate=True` imposes significantly higher memory requirements and shows much less benefit due to palatalization.

Instead, we should keep all contracted dimensions as size 1 in each chunk with `atop`, and then `sum` in another pass over all the dimensions with size 1 chunks. Instead of calling `einsum()` as the mapped function with `atop`, you would use a version of `einsum()` that keeps contracted dimensions as size one instead of removing them, like `keepdims=True` for `numpy.sum()`. The implementation would look somewhat similar to my implementation of `dask.array.isin` in #3363.

**Comment by shoyer** [comment]
Actually, it looks like @chrisbarber's latest version of `einsum` over in https://github.com/dask/dask/issues/732#issuecomment-369991557 already uses my suggestion above to avoid `concatenate=True`.

**Comment by sjperkins** [comment]
I'll take a look at implementing your review comments over the next couple of days.

**Comment by sjperkins** [comment]
Changes seemed fairly easy to make. Would you mind taking a look?

File: dask/array/routines.py [hunk 2]
```diff
@@ -1190,3 +1190,101 @@ def insert(arr, obj, values, axis):
...
+    # for the contracted dimensions
+    return chunk.reshape(chunk.shape + (1,) * kwargs['ncontract_inds'])
+
+
+def _einsum_kernel_optimize(*operands, **kwargs):
+    chunk = np.einsum(kwargs['subscripts'], *operands,
+                      casting=kwargs['casting'],
+                      dtype=kwargs['kernel_dtype'],
+                      order=kwargs['order'],
+                      optimize=kwargs['optimize'])
```
Comment (jakirkham) [inline dask/array/routines.py]: Might be able to `pop` a few values from `kwargs` and pass the rest through. That could also smooth over the differences between the supported optimize case and non-optimize/unsupported case. Thus simplifying the surrounding code.

Comment (shoyer) [reply dask/array/routines.py]: +1 for using `pop`, e.g.,
```
def _einsum_kernel(*operands, **kwargs):
    subscripts = kwargs.pop('subscripts')
    ncontract_inds = kwargs.pop('ncontract_inds')
    chunk = np.einsum(subsripts, *operands, **kwargs)
    ...
```

File: dask/array/routines.py [hunk 3]
```diff
@@ -1190,3 +1190,101 @@ def insert(arr, obj, values, axis):
...
+    # Avoid concatenate=True in atop by adding 1's
+    # for the contracted dimensions
+    return chunk.reshape(chunk.shape + (1,) * kwargs['ncontract_inds'])
+
+
+@wraps(np.einsum)
+def einsum(subscripts, *operands, **kwargs):
+    casting = kwargs.get('casting', 'safe')
+    dtype = kwargs.get('dtype')
+    optimize = kwargs.get('optimize', True)
```
Comment (shoyer) [inline dask/array/routines.py]: I was thinking something like:
```
optimize = kwargs.get('optimize')

...

if optimize is None:
    optimize = can_optimize
```

File: dask/array/routines.py [hunk 4]
```diff
@@ -1190,3 +1190,101 @@ def insert(arr, obj, values, axis):
...
+                         "of operands (%d)."
+                         % len(inputs), len(operands))
+
+    # Set of all indices
+    all_inds_set = set().union(*inputs)
+
+    # If output isn't provided, it's equal to
+    # the sorted list of all indices
+    if output_str is None:
+        output_str = list(sorted(all_inds_set))
```
Comment (shoyer) [inline dask/array/routines.py]: I thought repeated indices are summed over by default?

File: dask/array/tests/test_routines.py [hunk 5]
```diff
@@ -1286,3 +1286,37 @@ def test_result_type():
     # dask 0d-arrays are NOT treated like scalars
     c = da.from_array(np.ones((), np.float64), chunks=())
     assert da.result_type(a, c) == np.float64
+
+
+@pytest.mark.parametrize('einsum_signature', [
+    'abc,bad->abcd',
+    'abcdef,bcdfg->abcdeg',
+    'ea,fb,abcd,gc,hd->efgh',
```
Comment (shoyer) [inline dask/array/tests/test_routines.py]: it would be good to add a few more examples for more test coverage:
- with outputs omitted, e.g., `'ab,b'`
- with no output indices, e.g., `'ab,b->'
- `...` for broadcasting, e.g., `'...a,a'` (could raise an informative error for now)
- with the alternative syntax `einsum(op0, sublist0, op1, sublist1, ..., [sublistout])` (could raise an informative error for now)

**Comment by sjperkins** [comment]
Please note the changes I've made in https://github.com/dask/dask/pull/3412/commits/89c3979b0a2097ca50ba67f6388c1b47c2fecf11. If no output subscripts are provided, a considerable amount of logic would need to be coded into da.einsum in order to infer the output subscripts (and atop dimensions).

However, we can extract this information from `np.einsum_path` which outputs a complete subscript specification, including the output subscripts.

tl;dr if you have numpy < 1.12.0, you need to fully specify your output subscripts. I think this is a reasonable solution and fixes a bunch of xfails.

**Comment by sjperkins** [comment]
Looks like `six` isn't installed on the 3.6 travis? 

Snippet from the [log](https://api.travis-ci.org/v3/job/367683936/log.txt)

```

    $ conda update -n base conda



numpy 1.14.2:   0% 0/1 [00:00<?, ?it/s]
numpy 1.14.2:   0% 0.0012973293470298992/1 [00:00<02:29, 149.41s/it]
numpy 1.14.2:  62% 0.6227180865743516/1 [00:00<00:39, 104.64s/it]   
numpy 1.14.2:  81% 0.8092885375494071/1 [00:01<00:14, 75.96s/it] 
numpy 1.14.2:  94% 0.9433465085638999/1 [00:02<00:03, 53.77s/it]
numpy 1.14.2: 100% 1.0/1 [00:03<00:00,  3.83s/it]               
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/home/travis/build/dask/dask/dask/array/__init__.py", line 8, in <module>
    from .routines import (take, choose, argwhere, where, coarsen, insert,
  File "/home/travis/build/dask/dask/dask/array/routines.py", line 12, in <module>
    from six import string_types
ModuleNotFoundError: No module named 'six'
Array import failed
```

File: dask/array/routines.py [hunk 6]
```diff
@@ -9,6 +9,7 @@
 from operator import getitem
 
 import numpy as np
+from six import string_types
```
Comment (mrocklin) [inline dask/array/routines.py]: We actually don't depend on six.  I recommend just using `str` instead.  If you need something cross version then use `from .compatibility import str`

File: dask/array/routines.py [hunk 7]
```diff
@@ -1190,3 +1191,118 @@ def insert(arr, obj, values, axis):
...
+                                             optimize=optimize)
+
+        # We can obtain a more complete subscript specification
+        # from the path_info string, as it will fill in
+        # missing outputs and replace ellipses with concrete
+        # subscripts.
+        contraction_str = 'Complete contraction:'
+        start = path_info.find(contraction_str) + len(contraction_str)
+        end = path_info.find('\n')
+        subscripts = path_info[start:end].strip()
```
Comment (shoyer) [inline dask/array/routines.py]: Honestly, I'd rather require the more explicit syntax than parse this from a string, which seems quite fragile.

The other option is to copy `_parse_einsum_input()` from NumPy, which conveniently has the same license as dask (yay open source!):
https://github.com/numpy/numpy/blob/8323be1bc44c2811fc36f5b99c1a30ebcee8edbd/numpy/core/einsumfunc.py#L374

I think that function could work basically unchanged for dask, with `asanyarray` referring to the dask.array version (or maybe removed if unnecessary).

Comment (sjperkins) [reply dask/array/routines.py]: I was loathe to let go of the full functionality so I've gone with the `_parse_einsum_input()` approach.

File: dask/array/routines.py [hunk 8]
```diff
@@ -1190,3 +1192,74 @@ def insert(arr, obj, values, axis):
...
+@wraps(np.einsum)
+def einsum(*operands, **kwargs):
+    casting = kwargs.get('casting', 'safe')
+    dtype = kwargs.get('dtype')
+    optimize = kwargs.get('optimize')
+    order = kwargs.get('order', 'K')
+    einsum_dtype = dtype
+
+    inputs, outputs, ops = parse_einsum_input(operands)
+    subscripts = '->'.join((inputs,outputs))
```
Comment (shoyer) [inline dask/array/routines.py]: there are extra parentheses here

File: dask/array/einsum.py [hunk 9]
```diff
@@ -0,0 +1,171 @@
+from numpy.compat import basestring
+
+einsum_symbols = 'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ'
+einsum_symbols_set = set(einsum_symbols)
+
+
+def parse_einsum_input(operands):
```
Comment (shoyer) [inline dask/array/einsum.py]: please add a note about the source of this function, along the lines of what we see here:
https://github.com/dask/dask/blob/cab9b17174ca61b052f46fd735fa02edb5dc450d/dask/array/numpy_compat.py#L61-L65

**Comment by shoyer** [comment]
The other concern I can think of is operations on mixed numpy/dask arrays. Most dask.array functions support that, so it would be good to do so here as well. I think that would be as simple as using `dask.array.asarray()` in `parse_einsum_input()`.

File: dask/array/routines.py (hunk 8 above)
Comment (sjperkins) [reply dask/array/routines.py]: Those are necessary (`os.path.join` behaves differently, which is frustrating at times)

```python
>>> '->'.join(('foo', 'qux'))
'foo->qux'
```

File: dask/array/einsum.py (hunk 9 above)
Comment (sjperkins) [reply dask/array/einsum.py]: done

File: dask/array/routines.py (hunk 8 above)
Comment (shoyer) [reply dask/array/routines.py]: Oops, I missed the comma.

In that case, please add a space after the comma, per pep8!

Comment (sjperkins) [reply dask/array/routines.py]: Done

**Comment by sjperkins** [comment]
With regard to the trick where `concatenate=True` is avoided by:

- contraction dimensions are added to the end of the output dimensions
- Singleton axes for the contraction dimensions are added via reshape of the kernel result
- the result of atop is reduced over the contraction dims added to the output dims

1. What's the tradeoff like between avoiding the extra memory allocations required by using concatenate and adding the extra reduction keys in the graph?
2. Are the extra reduction keys inline/optimised away when the graph is executed?
3. I wonder if it would be possible to automate this, although a `sum` might not be the appropriate reduction operation (I see @shoyer uses `any` for `isin`)

File: dask/array/tests/test_routines.py (hunk 5 above)
Comment (dgasmith) [reply dask/array/tests/test_routines.py]: As a note I would also explicitly test broadcasting operations such as 
```
a.shape = (4, 1)
b.shape = (4, 4)
np.einsum('ij,jk->ik', a, b)
```

A few more similar tests can be found [here](https://github.com/dgasmith/opt_einsum/blob/master/opt_einsum/tests/test_contract.py#L247).

Einsums functionality is incredibly broad and I got caught on a few issues that were not tested in NumPy and as such NumPy 14.0 had a few regressions from my changes. A few other tricky cases I would recommend:

```
'aa->a`,
'ab,ab,c->c'   
'aab,bc->ac',
'aab,bcc->ac',
'fdf,cdd,ccd,afe->ae',
'fff,fae,bef,def->abd',
```

Comment (sjperkins) [reply dask/array/tests/test_routines.py]: @dgasmith OK, I've added the test cases in your comment as well as the broadcasting test cases. They succeed locally for me and I've pushed them up.

I realise that the range of einsum's behaviour is broad, but my available time for working on this PR is running out.

With regard to the dask wrapper,  as long as the full subscript specificaiton is present, dask will align each chunk correctly and defer to `np.einsum`'s implementation

**Comment by mrocklin** [comment]
> What's the tradeoff like between avoiding the extra memory allocations required by using concatenate and adding the extra reduction keys in the graph?

I'm coming into this conversation a little late so I may not fully understand the conversation so far.  However let me pose a question:

```python
x = da.arange(int(1e10), chunks=int(1e7))
da.einsum('a->', x)
```

Does this sum each of the chunks first and then concatenate those results and sum them?  Or does it concatenate first to a single 10GB array in memory and then sum?

If it does the latter then the concern isn't task overhead, it'sa question about if it is feasible to compute the result.

**Comment by sjperkins** [comment]
@mrocklin  With regards to your example:

 ```python
x = da.arange(int(1e10), chunks=int(1e7))
da.einsum('a->', x)
```
> Does this sum each of the chunks first and then concatenate those results and sum them? Or does it concatenate first to a single 10GB array in memory and then sum?

The above would perform an `np.einsum` in each chunk individually, producing scalars which would then be summed to produce the final scalar result.

With regard to https://github.com/dask/dask/pull/3412#issuecomment-382359464, I'm talking about

- contraction dims added to end of output dims [line](https://github.com/dask/dask/pull/3412/files#diff-02bfb8d5198c7e872d9e73af248a17a6R1256)
- contraction dims adjusted to size 1 [line](https://github.com/dask/dask/pull/3412/files#diff-02bfb8d5198c7e872d9e73af248a17a6R1247)
- output of kernel reshaped to add size 1 dims for each contraction dim [line](https://github.com/dask/dask/pull/3412/files#diff-02bfb8d5198c7e872d9e73af248a17a6R1205)
- final reduction of size 1 contraction dims [line](https://github.com/dask/dask/pull/3412/files#diff-02bfb8d5198c7e872d9e73af248a17a6R1263)

This is what @shoyer did in his `isin` implementation and it's a neat way of getting around the the singleton lists passed into the kernels by atop in cases of atop dim contraction. Quote from atop docs

> Any index, like i missing from the output index is interpreted as a contraction (note that this differs from Einstein convention; repeated indices do not imply contraction.) In the case of a contraction the passed function should expect an iterable of blocks on any array that holds that index. To receive arrays concatenated along contracted dimensions instead pass concatenate=True.

**Comment by sjperkins** [comment]
>>    Does this sum each of the chunks first and then concatenate those results and sum them? Or does it concatenate first to a single 10GB array in memory and then sum?

> The above would perform an np.einsum in each chunk individually, producing scalars which would then be summed to produce the final scalar result.

In fact, this is what's so neat about dask einsum: as long as we have fully specified subscripts (now produced by parse_einsum_input), dask atop produces the correct broadcasting. I think `parse_einsum_input` is larger than the actual `da.einsum` implementation.

**Comment by mrocklin** [comment]
OK, yeah cool.  I should have just tried that directly (which I just did and was happy with the results)

File: dask/array/routines.py [hunk 10]
```diff
@@ -1190,3 +1192,74 @@ def insert(arr, obj, values, axis):
...
+    result = atop(_einsum_kernel, tuple(outputs) + tuple(contract_inds),
+                  *(a for ap in zip(ops, inputs) for a in ap),
+                  **atop_kwargs)
+
+    # Now reduce over any extra contraction dimensions
+    if ncontract_inds > 0:
+        size = len(outputs)
+        result = result.sum(axis=list(range(size, size + ncontract_inds)))
+
+    return result
```
Comment (mrocklin) [inline dask/array/routines.py]: It seems odd to me that some of the einsum functions in a separate einsum.py files but not others?  Should we consolidate these either in einsum.py or routines.py?

File: dask/array/tests/test_routines.py [hunk 11]
```diff
@@ -1286,3 +1286,121 @@ def test_result_type():
...
+                                  .split(','))
+
+    shapes = [shape_from_string(s) for s in input_sigs]
+    chunks = [chunks_from_string(s) for s in input_sigs]
+
+    np_inputs = [np.random.random(s) for s in shapes]
+    da_inputs = [da.from_array(i, chunks=c) for i, c in zip(np_inputs, chunks)]
+
+    assert_eq(np.einsum(einsum_signature, *np_inputs),
+              da.einsum(einsum_signature, *da_inputs))
```
Comment (mrocklin) [inline dask/array/tests/test_routines.py]: Nice test!  Thank you for writing this.

File: dask/array/tests/test_routines.py [hunk 12]
```diff
@@ -1286,3 +1286,121 @@ def test_result_type():
...
+
+
+def test_einsum_broadcasting_contraction3():
+    a = np.random.rand(1, 5, 4)
+    b = np.random.rand(4, 1, 6)
+    c = np.random.rand(5, 6)
+    d = np.random.rand(7, 7)
+
+    d_a = da.from_array(a, chunks=(1,(2,3),(2,2)))
+    d_b = da.from_array(b, chunks=((2,2), 1,  (4,2)))
```
Comment (mrocklin) [inline dask/array/tests/test_routines.py]: I'm somewhat surprised that flake8 isn't yelling at you here.  Typically it requires exactly one space after every comma.

File: dask/array/routines.py (hunk 10 above)
Comment (sjperkins) [reply dask/array/routines.py]: Placed `einsum` function in `einsum.py` and imported it in `routines.py`

File: dask/array/tests/test_routines.py (hunk 12 above)
Comment (sjperkins) [reply dask/array/tests/test_routines.py]: Fixed

**Comment by mrocklin** [comment]
I'm not accurately tracking the progress of the conversation between
yourself and @shoyer.  Have the comments that he has raised above been
addressed?  @shoyer, do you have other concerns about this?

**Comment by sjperkins** [comment]
> Have the comments that he has raised above been addressed?

I believe so. Regarding his not-in-the-code comments:

- https://github.com/dask/dask/pull/3412#issuecomment-382128268 fixed by https://github.com/dask/dask/pull/3412/commits/9710c7bffff69018b6295f1fd7c19b2e9494b028
- https://github.com/dask/dask/pull/3412#issuecomment-382128701 fixed by https://github.com/dask/dask/pull/3412/commits/fcaf96d3ecf80f870ee6acfb0d565636b988ba83

**Comment by mrocklin** [comment]
@shoyer anything further from you.  Are you +1 ?

File: dask/array/tests/test_routines.py [hunk 13]
```diff
@@ -1286,3 +1286,121 @@ def test_result_type():
...
+    chunks = [chunks_from_string(s) for s in input_sigs]
+
+    np_inputs = [np.random.random(s) for s in shapes]
+    da_inputs = [da.from_array(i, chunks=c) for i, c in zip(np_inputs, chunks)]
+
+    assert_eq(np.einsum(einsum_signature, *np_inputs),
+              da.einsum(einsum_signature, *da_inputs))
+
+
+def test_einsum_broadcasting_contraction():
```
Comment (shoyer) [inline dask/array/tests/test_routines.py]: Nit: I'm not quite sure why you call this test "broadcasting"? Based on the example, I don't see any broadcasting -- there's no use of `'...'` and all the dimension sizes match. So maybe better to remove that part of the name (here and below).

**Comment by shoyer** [comment]
I have one more nit on the test names, but I've resolved all my substantiative concerns. So +1 from me too -- thanks @sjperkins for putting together this very useful function!

File: dask/array/tests/test_routines.py (hunk 13 above)
Comment (sjperkins) [reply dask/array/tests/test_routines.py]: Ah good catch. The test cases were adapted from [here](https://github.com/dgasmith/opt_einsum/blob/master/opt_einsum/tests/test_contract.py#L247) as suggested by https://github.com/dask/dask/pull/3412#discussion_r182413444.

I've reintroduced the multiplication terms that tested the broadcast.

**Comment by shoyer** [comment]
> I wonder if it would be possible to automate this, although a sum might not be the appropriate reduction operation (I see @shoyer uses any for isin)

I agree, this is definitely worth considering (in a follow-up PR, of course). We might appropriate call this pattern "map reduce", since it involves a map step followed by a reduction, and in many ways works quite similarly to classic map-reduce algorithms (even with tree-reduction vs. combiners).

So perhaps `dask.array.mapreduce` is order? Or I suppose we could squeeze this functionality (inserting dummy dimensions/chunks and a reduction) into `atop` by adding a `reduce` keyword argument (e.g., `reduce=da.sum` for this example).

**Comment by shoyer** [comment]
> Would it not be possible to simply rewrite the keys of the result of atop to remove the extra (singleton) dims?

Sorry, I'm not quite sure what you're suggesting here. If you're concerned about the overhead of mapping another function over each chunk to insert dummy dimensions, then I agree that that should not be necessary (in theory) with customized reduction logic.

**Comment by sjperkins** [comment]
> Sorry, I'm not quite sure what you're suggesting here. If you're concerned about the overhead of mapping another function over each chunk to insert dummy dimensions, then I agree that that should not be necessary (in theory) with customized reduction logic.

The process described in the second half of https://github.com/dask/dask/pull/3412#issuecomment-382399633, requires an extra `da.sum` (for `einsum`) or `da.any` (for `isin`) to *remove* the dummy singleton contraction dimensions. I think this introduces an extra layer of tasks into the graph that could probably be elided by additional logic in **top** or **atop**.

File: dask/array/einsum.py [hunk 14]
```diff
@@ -0,0 +1,253 @@
...
+from .core import (atop, asarray)
+
+einsum_symbols = 'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ'
+einsum_symbols_set = set(einsum_symbols)
+
+
+# This function duplicates numpy's _parse_einsum_input() function
+# See https://github.com/numpy/numpy/blob/master/LICENSE.txt
+# or NUMPY_LICENSE.txt within this directory
+def parse_einsum_input(operands):
```
Comment (jakirkham) [inline dask/array/einsum.py]: Just to help me understand, was this added here for compat with old versions of NumPy or was this added to workaround conversion (in the original code) of objects to NumPy arrays instead of Dask arrays?

**Comment by jakirkham** [comment]
I may have overlooked this, but could we add some tests for `optimize` (assuming they don't already exist)?

File: dask/array/einsum.py (hunk 14 above)
Comment (sjperkins) [reply dask/array/einsum.py]: @jakirkham `np.einsum` subscript specifications do not necessarily require the output subscripts to be specified and can have ellipses to indicate broadcasting. `parse_einsum_input` replaces and introduces subscripts for ellipses and outputs respectively, allowing the dask einsum wrapper to properly broadcast chunks. Without this, the logic for `np.einsum` functionality would need to be written in the dask wrapper.

**Comment by sjperkins** [comment]
> I may have overlooked this, but could we add some tests for optimize (assuming they don't already exist)?

You did not, there no specific tests for `optimize`. Unfortunately I've reached the time limit available for this PR.

Perhaps this can be addressed in a follow-up PR?

**Comment by shoyer** [comment]
I don't think we need tests for `optimize`. Dask's `einsum` isn't doing anything beyond passing the optimize argument on to NumPy.

**Comment by jakirkham** [comment]
Right, my only interest was in seeing that it *does* get passed on. This could be a very simple test IMHO.

**Comment by shoyer** [comment]
Tests are failing with a lint error:
```
dask/array/routines.py:26:1: F401 '.einsum.einsum' imported but unused
```

(Run `flake8 dask` locally to test this)

File: dask/array/einsum.py (hunk 14 above)
Comment (jakirkham) [reply dask/array/einsum.py]: Sorry I haven't been following this PR very closely. I thought this code was lifted from an existing (perhaps internal) NumPy function. Is that not true?

Comment (shoyer) [reply dask/array/einsum.py]: That's exactly right, this was lifted from an internal NumPy function. It handles all of einsum's messy input parsing (broadcasting, optional outputs, multiple argument signatures), which lets us write a really simple implementation for dask.

**Comment by shoyer** [comment]
> Right, my only interest was in seeing that it does get passed on. This could be a very simple test IMHO.

I'm not sure how we would test this short of mocking NumPy's einsum, which seems like overkill.

I guess we could include something super simple, e.g., to verify that `optimize=True` and `optimize=False` give the same result.

**Comment by jakirkham** [comment]
Right was only thinking of comparing NumPy and Dask when given the same `optimize` argument.

File: dask/array/einsum.py (hunk 14 above)
Comment (jakirkham) [reply dask/array/einsum.py]: Was it lifted because of the `asarray` call or is there more going on in this copy-paste?

Comment (shoyer) [reply dask/array/einsum.py]: It's a private function `_parse_einsum_input()` in NumPy -- not something that we could rely on reliably without forking.

**Comment by sjperkins** [comment]
> Right, my only interest was in seeing that it does get passed on. This could be a very simple test IMHO.

Apologies and good catch, this revealed a mistake in how `np.einsum_path` was called. I've fixed it and added some permuted optimize test cases.  Hope this brings the PR up to speed.

**Comment by sjperkins** [comment]
Rats, just thought of something ... in order to produce an optimized path, `np.einsum_path` must be called (either in da.einsum or outside) with dask Arrays, which probably get converted to numpy arrays via internal asarray calls.

Any thoughts on how to work around this?

**Comment by jakirkham** [comment]
Copy it too? Maybe someone else has a better idea.

**Comment by jakirkham** [comment]
Thanks for the optimization tests. Looks good.

File: dask/array/einsum.py (hunk 14 above)
Comment (jakirkham) [reply dask/array/einsum.py]: Do you think we could encourage NumPy to make it public/semi-public somehow or are there some obstacles to that?

Should add this is not a blocker. Just thinking of reducing our maintenance burden over the long run.

Comment (sjperkins) [reply dask/array/einsum.py]: > Do you think we could encourage NumPy to make it public/semi-public somehow or are there some obstacles to that?

I reckon thats a good idea. @dgasmith might have some thoughts on this?

**Comment by shoyer** [comment]
> Any thoughts on how to work around this?

You can make dummy numpy arrays with the right shape with repeated values (with strides) by using `np.broadcast_to`, e.g., `np.broadcast_to(0, shape=(1000, 2000, 3000))`.

**Comment by sjperkins** [comment]
> You can make dummy numpy arrays with the right shape with repeated values (with strides) by using np.broadcast_to, e.g., np.broadcast_to(0, shape=(1000, 2000, 3000)) actually only uses a few bytes to store its data, not many gigabytes.

Oh nice. I actually thought of [np.empty((1,)*o.ndim, o.dtype) for o in operands] but this should work well!

**Comment by shoyer** [comment]
Yeah, it's important to give einsum_path the right shapes in order for the cost to be estimated correctly.

**Comment by sjperkins** [comment]
OK I hope thats it. tests and flake8 succeed locally for me.

**Comment by shoyer** [comment]
Travis-CI is failing due to an ImportError for toolz in unrelated files:
https://travis-ci.org/dask/dask/jobs/368350028

This looks entirely unrelated, so I think we are good to merge here.

> The process described in the second half of #3412 (comment), requires an extra da.sum (for einsum) or da.any (for isin) to remove the dummy singleton contraction dimensions. I think this introduces an extra layer of tasks into the graph that could probably be elided by additional logic in top or atop, by simply reducing the dimensionality of the appropriate key tuples.

I don't think this is possible. The extra summation is necessary for combining the chunks, whether they have size 1 or are scalars.

Suppose we want to sum an array with 1000 chunks of size 1000 each, e.g., `da.arange(int(1e6), chunks=1000).sum()` using einsum:
1. We need to sum each chunk. Now we have 1000 intermediate results that are either size `(1,)` or scalars.
2. We need to sum the intermediate results, by adding them all together.

There is no skipping step 2. The individual chunks need to be combined somehow.

**Comment by mrocklin** [comment]
OK, merging.  Thank you so much @sjperkins for the great work here.  Thank you @shoyer, @jakirkham, and @dgasmith for the careful review

**Comment by sjperkins** [comment]
Likewise, thanks for the reviews, assistance and accepting the PR.

Also thanks to @chrisbarber for providing the base implementation.

**Comment by jakirkham** [comment]
Thanks @sjperkins. Very nice contribution. :)

**Comment by jakirkham** [comment]
FYI from `master` on Python 3.4, NumPy 1.10.

Side note: Sorry for the junk in the snippet. Travis CI injects some term color junk in there that wasn't easy to strip.

```python
[1m[31m_____________________ test_einsum_optimize[optimize_opts0] _____________________[0m
[gw2] linux -- Python 3.4.5 /home/travis/miniconda/envs/test-environment/bin/python

optimize_opts = (True, False)

[1m    @pytest.mark.parametrize('optimize_opts', [[0m
[1m        (True, False),[0m
[1m        ('greedy', False),[0m
[1m        ('optimal', False)[0m
[1m    ])[0m
[1m    def test_einsum_optimize(optimize_opts):[0m
[1m        sig = 'ea,fb,abcd,gc,hd->efgh'[0m
[1m        input_sigs = sig.split('->')[0].split(',')[0m
[1m        np_inputs, da_inputs = _numpy_and_dask_inputs(input_sigs)[0m
[1m    [0m
[1m        opt1, opt2 = optimize_opts[0m
[1m    [0m
[1m>       assert_eq(np.einsum(sig, *np_inputs, optimize=opt1),[0m
[1m                  da.einsum(sig, *np_inputs, optimize=opt2))[0m
[1m[31mE       TypeError: 'optimize' is an invalid keyword for einsum[0m

[1m[31mdask/array/tests/test_routines.py[0m:1371: TypeError
[1m[31m_____________________ test_einsum_optimize[optimize_opts1] _____________________[0m
[gw2] linux -- Python 3.4.5 /home/travis/miniconda/envs/test-environment/bin/python

optimize_opts = ('greedy', False)

[1m    @pytest.mark.parametrize('optimize_opts', [[0m
[1m        (True, False),[0m
[1m        ('greedy', False),[0m
[1m        ('optimal', False)[0m
[1m    ])[0m
[1m    def test_einsum_optimize(optimize_opts):[0m
[1m        sig = 'ea,fb,abcd,gc,hd->efgh'[0m
[1m        input_sigs = sig.split('->')[0].split(',')[0m
[1m        np_inputs, da_inputs = _numpy_and_dask_inputs(input_sigs)[0m
[1m    [0m
[1m        opt1, opt2 = optimize_opts[0m
[1m    [0m
[1m>       assert_eq(np.einsum(sig, *np_inputs, optimize=opt1),[0m
[1m                  da.einsum(sig, *np_inputs, optimize=opt2))[0m
[1m[31mE       TypeError: 'optimize' is an invalid keyword for einsum[0m

[1m[31mdask/array/tests/test_routines.py[0m:1371: TypeError
[1m[31m_____________________ test_einsum_optimize[optimize_opts2] _____________________[0m
[gw2] linux -- Python 3.4.5 /home/travis/miniconda/envs/test-environment/bin/python

optimize_opts = ('optimal', False)

[1m    @pytest.mark.parametrize('optimize_opts', [[0m
[1m        (True, False),[0m
[1m        ('greedy', False),[0m
[1m        ('optimal', False)[0m
[1m    ])[0m
[1m    def test_einsum_optimize(optimize_opts):[0m
[1m        sig = 'ea,fb,abcd,gc,hd->efgh'[0m
[1m        input_sigs = sig.split('->')[0].split(',')[0m
[1m        np_inputs, da_inputs = _numpy_and_dask_inputs(input_sigs)[0m
[1m    [0m
[1m        opt1, opt2 = optimize_opts[0m
[1m    [0m
[1m>       assert_eq(np.einsum(sig, *np_inputs, optimize=opt1),[0m
[1m                  da.einsum(sig, *np_inputs, optimize=opt2))[0m
[1m[31mE       TypeError: 'optimize' is an invalid keyword for einsum[0m

[1m[31mdask/array/tests/test_routines.py[0m:1371: TypeError
```

ref: https://travis-ci.org/dask/dask/jobs/369157407
ref: https://api.travis-ci.org/v3/job/369157407/log.txt

**Comment by shoyer** [comment]
Oops, I guess we need to add a skipif.

**Comment by jakirkham** [comment]
Resolved with PR ( https://github.com/dask/dask/pull/3428 ).
